File: selected_coding_interview/codes/java/lc_238_product_of_array_except_self/lc_238_product_of_array_except_self.java
/*
* File: lc_238_product_of_array_except_self.java
* Created Time: 2025-12-30
* Author: krahets
*/

package lc_238_product_of_array_except_self;

import include.*;
import java.util.*;

// ===== Solution Code =====
    class Solution {
        public int[] productExceptSelf(int[] nums) {
            int len = nums.length;
            if (len == 0) return new int[0];
            int[] ans = new int[len];
            ans[0] = 1;
            int tmp = 1;
            for (int i = 1; i < len; i++) {
                ans[i] = ans[i - 1] * nums[i - 1];
            }
            for (int i = len - 2; i >= 0; i--) {
                tmp *= nums[i + 1];
                ans[i] *= tmp;
            }
            return ans;
        }
    }

public class lc_238_product_of_array_except_self {
    public static void main(String[] args) {
        // ======= Test Case =======
        // Test case 1
        int[] test_input = new int[]{1, 2, 3, 4, 5};

        // ====== Driver Code ======
        Solution slt = new Solution();
        var result = slt.productExceptSelf(test_input);
        System.out.println(result);

    }
}

File: selected_coding_interview/codes/python/lc_238_product_of_array_except_self.py
"""
File: lc_238_product_of_array_except_self.py
Created Time: 2025-12-30
Author: krahets
"""

from include import *


# ===== Solution Code =====
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        ans, tmp = [1] * len(nums), 1
        for i in range(1, len(nums)):
            ans[i] = ans[i - 1] * nums[i - 1] # 下三角
        for i in range(len(nums) - 2, -1, -1):
            tmp *= nums[i + 1]                # 上三角
            ans[i] *= tmp                     # 下三角 * 上三角
        return ans


# ======= Test Case =======
# Test case 1
test_input = [1, 2, 3, 4, 5]

# ====== Driver Code ======
slt = Solution()
result = slt.productExceptSelf(test_input)
print(result)
